### ISSUE TITLE
use of GPS info in openMVG

### ISSUE THREAD

**Issue by ywjia** [openMVG/openMVG#822]
Hi, I have a question about using GPS prior (from exif) in openMVG. 

I understand from the document that the syntax / pipeline of using GPS prior should look like the following (only listed relevant parameters):

(1) openMVG_main_SfMInit_ImageListing -P -W w0 w1 w2 ...
(2) openMVG_main_ComputeFeatures  ...
(3) openMVG_main_ListMatchingPairs -G -n limit_num_of_pairs -o match_dir/pair_list.txt ...
(4) openMVG_main_ComputeMatches -l match_dir/pair_list.txt ...
(5) openMVG_main_IncrementalSfM -P ...

After that, I can use the following to generate dense point cloud.

(6) openMVG_main_openMVG2openMVS -i recon_dir/sfm_data.bin -o recon_dir/scene.mvs -d ./undisto
(7) DensifyPointCloud -i recon_dir/scene.mvs -o recondir/scene_dense.ply

In comparing with the dense point cloud generated without using GPS prior, the output from the above process is noticeably sparser. I understand that by using GPS prior, the process is constrained and hence fewer points. However, even if I set w0=w1=w2=0.0 (which supposes to give 0 weight to GPS prior), and set limit_num_of_pairs to a large number (say, 1024, more than the number of images), the result stays the same. This really puzzles me since I thought the above setting should be equivalent to not using GPS prior at all.

Moreover, the dense point cloud generated using GPS prior contains some strange patterns of lines, and after subsequent processing (ReconstructMesh and TextureMesh), the result contains significantly less details.

Any ideas of what the issue is, please?

Thanks.

**Comment by pmoulon** [comment]
@ywjia Dense reconstruction process does not like to have large coordinates.
Since the SfM reconstruction is in the geodesic coordinate system and so use large doubles coordinates, you can see some lines appears due to the quantization (double to float conversion done in some third party densification project).

It's often better to convert back the SfM reconstruction to an unit norm scale and center it on 0,0,0 and then rescale it to the geodesic coordinates once the dense reconstruction is done.

Regarding the GPS weight == 0 => In the way the current GPS registration is done, setting the weight to 0 is not sufficient, there is still a rigid registration that is done before using GPS constraint for a better BA conditioning.

Hoping it helps

**Comment by ywjia** [comment]
@pmoulon Thanks a lot for the explanation. It makes a lot of sense. Here is what I did according to your suggestion.

In "main_SfMInit_ImageListing.cpp", in the function "checkGPS( )", instead of calling "lla_to_ecef( )" to convert lla to ecef, I added a function to convert lla to local North-East-Up coordinates, using the first-view as the reference point. The issue seems to be resolved by this change. Actually, the densified point cloud contains a slightly more points, and the strange pattern of lines is completely gone,
 
Please let me know if the above makes sense.
Thanks again.

**Comment by jpwhitney** [comment]
@ywjia This method sounds great, can you share the function?

**Comment by hcjghr** [comment]
If I can suggest an option that you can also specify the reference point. And for example if its not specified the first camera is used. The reason is that the reference point is often defined prior to the mission and can be used by multiple applications/people in different area so its important to have a way to enforce it.

**Comment by jpwhitney** [comment]
@hcjghr That sounds useful.  Thank you for your comments.  How does one specify the reference point?  And at which stage of the SfM pipeline?

**Comment by ywjia** [comment]
@jpwhitney Here is the function to convert lla to neu:

Vec3 lla_to_neu (double lat, double lon, double alt, double lat_ref, double lon_ref, double alt_ref)
{
  const double clat = cos( D2R(lat) );
  const double slat = sin( D2R(lat) );
  const double clon = cos( D2R(lon) );
  const double slon = sin( D2R(lon) );

  const double a2 = Square(WGS84_A);
  const double b2 = Square(WGS84_B);

  const double L = 1.0 / sqrt(a2 * Square(clat) + b2 * Square(slat));
  const double x = (a2 * L + alt) * clat * clon;
  const double y = (a2 * L + alt) * clat * slon;
  const double z = (b2 * L + alt) * slat;

  const double clat_ref = cos( D2R(lat_ref) );
  const double slat_ref = sin( D2R(lat_ref) );
  const double clon_ref = cos( D2R(lon_ref) );
  const double slon_ref = sin( D2R(lon_ref) );

  const double L_ref = 1.0 / sqrt(a2 * Square(clat_ref) + b2 * Square(slat_ref));
  const double x_ref = (a2 * L_ref + alt_ref) * clat_ref * clon_ref;
  const double y_ref = (a2 * L_ref + alt_ref) * clat_ref * slon_ref;
  const double z_ref = (b2 * L_ref + alt_ref) * slat_ref;

  const double dx = x - x_ref;
  const double dy = y - y_ref;
  const double dz = z - z_ref;

  const double e = -slon_ref * dx + clon_ref * dy;
  const double n = -slat_ref * clon_ref * dx - slat_ref * slon_ref * dy + clat_ref * dz;
  const double u = clat_ref * clon_ref * dx + clat_ref * slon_ref * dy + slat_ref * dz;

  return Vec3(e, n, u);
}

**Comment by ywjia** [comment]
@hcjghr My first thought was to add a user-specified reference point, but this would require some interface changes. So I decided to try the simplest method of using the first camera as the reference. Another way is to find the camera at the center of the cameras and use it as the reference point.

**Comment by ywjia** [comment]
When I

**Comment by pmoulon** [comment]
@ywjia 
We could think about to add a tool to normalize a scene (according a reference image or just a global normalization). Any though about that?

We can also integrate your lla_to_neu transform.

**Comment by ywjia** [comment]
@pmoulon For applications of UAV imagery with GPS data, it seems to make sense to use ENU coordinate system, since in such applications, the areas covered are normally small relative to the earth radius and large relative to GPS accuracy.  

Not sure what exactly you meant by "normalize a scene".

**Comment by jpwhitney** [comment]
@ywjia I tried inserting the function you mentioned but I get a compile error.  I think it's because the lla_to_ecef( ) function that replaced by the lla_to_neu() function only has 3 arguments.

Any idea of what I did wrong?  I've included the edited version of main_SfMInit_ImageListing.cpp, renamed to .txt.
[main_SfMInit_ImageListing.txt](https://github.com/openMVG/openMVG/files/866733/main_SfMInit_ImageListing.txt)

**Comment by ywjia** [comment]
@jpwhitney You need to add lla_to_neu( ) as a separate function (in geodesy.hpp). You also need to initialize the three additional parameters (lat_ref, lon_ref, alt_ref) some way, for example, from the first camera.

**Comment by jpwhitney** [comment]
So within main_SfMInit_ImageListing.cpp there is the ability to obtain the initialization values for the three additional parameters (lat_ref, lon_ref, alt_ref)?

**Comment by ywjia** [comment]
@jpwhitney I should have used "first view" rather than "first camera".

To get (lat_ref, lon_ref, alt_ref), I simply did the following in main_SfMInit_ImageListing.cpp:

static int first_view = 1;
static double latitude_ref;
static double longitude_ref;
static double altitude_ref;

std::pair<bool, Vec3> checkGPS (  const std::string & filename) 
{
...
#if 0
        val.second = lla_to_ecef( latitude, longitude, altitude );
#else
	if (first_view==1) {
		latitude_ref = latitude;
		longitude_ref = longitude;
		altitude_ref = altitude;
		first_view = 0;
	}
	val.second = lla_to_neu( latitude, longitude, altitude, latitude_ref, longitude_ref, altitude_ref );
#endif
...
}

**Comment by hcjghr** [comment]
Hi again,

I remembered that I have a similar problem previously and I have created a binary that moved the origin of the whole scene to a user defined local origin. In my implementation binary **main_ChangeLocalOrigin** should be called with -l "x;y;z; of the local origin. This will create a new sfm_data_local.bin file which can be then used as any previous file. 

The code is available as a branch at: [https://github.com/hcjghr/openMVG/blob/develop_local_origin/]. Feel free to use it. @pmoulon if you think that would be useful I can add it as PR.
